bindings/go/constructor/construct.go

```diff
@@ -0,0 +1,499 @@
+	var err error
+
+	switch {
+	case resource.HasInput():
+		logger.Debug("processing resource with input method")
+		res, err = c.processResourceWithInput(ctx, targetRepo, resource, component, version)
+	case resource.HasAccess():
+		if byValue := c.opts.ProcessResourceByValue != nil && c.opts.ProcessResourceByValue(resource); byValue {
+			logger.Debug("processing resource by value")
+			res, err = c.processResourceByValue(ctx, targetRepo, resource, component, version)
+		} else {
+			logger.Debug("processing resource with existing access")
+			converted := constructorv1.ConvertToRuntimeResource(*resource)
+			res = &converted
+
+			if c.opts.ResourceDigestProcessorProvider != nil {
+				var digestProcessor ResourceDigestProcessor
+				if digestProcessor, err = c.opts.GetDigestProcessor(ctx, res); err == nil {
+					logger.Debug("processing resource digest")
+					if res, err = digestProcessor.ProcessResourceDigest(ctx, res); err != nil {
+						return nil, fmt.Errorf("error processing resource %q with digest processor: %w", resource.ToIdentity(), err)
+					}
+				}
+			}
+		}
+	default:
+		return nil, fmt.Errorf("resource %q has no access type and no input method", resource.ToIdentity())
+	}
+
+	if err != nil {
+		return nil, fmt.Errorf("error processing resource %q: %w", resource.ToIdentity(), err)
+	}
+
+	if res.Access == nil {
+		return nil, fmt.Errorf("after the input method was processed, no access was present in the resource. This is likely a problem in the input method")
+	}
+
+	logger.Debug("resource processed successfully")
+	return res, nil
+}
+
+func (c *DefaultConstructor) processResourceByValue(ctx context.Context, targetRepo TargetRepository, resource *constructorv1.Resource, component, version string) (*descriptor.Resource, error) {
+	repository, err := c.opts.GetResourceRepository(ctx, resource)
+	if err != nil {
+		return nil, err
+	}
+	converted := constructorv1.ConvertToRuntimeResource(*resource)
+	data, err := repository.DownloadResource(ctx, &converted)
+	if err != nil {
+		return nil, fmt.Errorf("error downloading resource: %w", err)
+	}
+	return addColocatedResourceLocalBlob(ctx, targetRepo, component, version, resource, data)
+}
+
+func (c *DefaultConstructor) processSource(ctx context.Context, targetRepo TargetRepository, src *constructorv1.Source, component, version string) (*descriptor.Source, error) {
+	logger := log.Base().With(
+		"component", component,
+		"version", version,
+		"source", src.ToIdentity(),
+	)
+	logger.Debug("processing source")
+
+	var res *descriptor.Source
+	var err error
+
+	if src.HasInput() {
+		logger.Debug("processing source with input method")
+		res, err = c.processSourceWithInput(ctx, targetRepo, src, component, version)
+	} else {
+		logger.Debug("processing source with existing access")
+		converted := constructorv1.ConvertToRuntimeSource(*src)
+		res = &converted
+	}
+
+	if err != nil {
+		return nil, fmt.Errorf("error processing source %q: %w", src.ToIdentity(), err)
+	}
```

Comment:
```suggestion
	if src.HasInput() {
		logger.Debug("processing source with input method")
		res, err = c.processSourceWithInput(ctx, targetRepo, src, component, version)
		if err != nil {
			return nil, fmt.Errorf("error processing source %q: %w", src.ToIdentity(), err)
		}
	} else {
		logger.Debug("processing source with existing access")
		converted := constructorv1.ConvertToRuntimeSource(*src)
		res = &converted
	}


```

Reply:
This is moot imo. I would argue that keeping nesting level low is preferrable always. But I also understand that not having the error present immediately hinders readability. WDYT?